Fix yaxis plot scaling

gjr80 opened the pull request:
`weeplot.utilities.scale()` returns unusable results when the data min/max values (`fmn`, `fmx`) are the same/close together and a lower bound (`minscale`) that is significantly different to the data values is specified. For example:
```
scale(100.0, 100.0, (0, None, None))
(0, 101.0, 0.2)
```
I suspect a similar issue will arise should an upper bound (`maxscale`) be specified.

The modified code now returns:
```
scale(100.0, 100.0, (0, None, None))
(0, 120.0, 20.0)
```
I have run the other `scale()` doctests through the modified code and they return the same results as the original code.

The `scale()` whilst not overly complex has a lot of corner cases to deal with so would appreciate another set of eyes (is this Matthew's code?) to check that I have not broken some other cases.

Oh, and I updated the doctests to python3 as well (print statements/functions)

matthewwall commented:
> The scale() whilst not overly complex has a lot of corner cases to deal with so would appreciate another set of eyes (is this Matthew's code?) to check that I have not broken some other cases.

i admit to chronic complexity, but blame says that tom is the most recent to modify that code

https://github.com/weewx/weewx/blame/master/bin/weeplot/utilities.py

tkeffer commented:
I don't think the doctests are entirely up to date (that's why they still use Python 2 `print` statements).

Put your new tests in `tests/test_utilities.py`.

gjr80 commented:
> I don't think the doctests are entirely up to date (that's why they still use Python 2 `print` statements).
> 
> Put your new tests in `tests/test_utilities.py`.

OK. So is there a simple way to reject one commit in a PR, I know it’s easy to add commits. Testing my git knowledge.

tkeffer commented:
If you just do another commit it will be combined with this PR. I think.

tkeffer commented:
Sorry. Not following. Why put things back as it was?

gjr80 commented:
> Sorry. Not following. Why put things back as it was?

The doctests, was going to put them back as they were.

> Put your new tests in `tests/test_utilities.py`.

Perhaps I misunderstood this, thought I needed to revert the doc tests and put the new tests in `tests/test_utilities.py`

tkeffer commented:
> The doctests, was going to put them back as they were.

No, leave them! It's a good thing that you ported them to Python 3. They are still useful as documentation for how the utilities work.

It's just that they are no longer the definitive tests. Those are in `tests/test_utilities.py`.

gjr80 commented:
Understand. Will leave this PR as is then. Will deal with `tests/test_utilities.py` separately.

tkeffer commented:
Anything new on this? I'd feel a lot better about accepting it if it passes the tests in `tests/test_utilities.py`.

gjr80 commented:
My apologies, will fix tomorrow.

gjr80 commented:
It all comes back to me now....I know why I have this natural aversion to working with the test suite....it's the journey, always seem to get to the destination in the end, it's just the journey is pretty crap.

Added another doctest and updated the test suite, as far as I can tell the `weeplot` utilities tests are working fine.

tkeffer commented:
Well, thousands of WeeWX fans worldwide thank you!

tkeffer commented:
Gary,

I just noticed that test suite `weeplot/tests/test_utilities.py` is now getting an error:

```shell
tkeffer@nuc:~/git/weewx/bin/weeplot/tests$ PYTHONPATH=../.. python test_utilities.py 
..F..
======================================================================
FAIL: test_scale (__main__.WeePlotUtilTest)
Test function scale()
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_utilities.py", line 42, in test_scale
    "(10.00000, 10.12000, 0.02000)")
AssertionError: '(10.00000, 10.10000, 0.01000)' != '(10.00000, 10.12000, 0.02000)'

----------------------------------------------------------------------
Ran 5 tests in 0.002s

FAILED (failures=1)
```

gjr80 commented:
Interesting, can’t see any external dependencies and my PR was the last change. Will have a look today.

gjr80 commented:
Tom,

What did you see, I just pulled down `master` and ran the test suite with no errors:
```
gary@stretch19:~/src/weewx$ make test
running bin/weecfg/tests/test_config.py
running bin/weeutil/tests/test_config.py
running bin/weeutil/tests/test_sun.py
running bin/weeutil/tests/test_weeutil.py
running bin/weewx/tests/test_units.py
running bin/weewx/tests/test_templates.py
running bin/weewx/tests/test_accum.py
running bin/weewx/tests/test_engine.py
running bin/weewx/tests/test_wxformulas.py
running bin/weewx/tests/test_wxservices.py
running bin/weewx/tests/test_cheetah.py
running bin/weewx/tests/test_restx.py
running bin/weewx/tests/test_almanac.py
running bin/weewx/tests/test_database.py
running bin/weewx/tests/test_aggregate.py
running bin/weewx/tests/test_daily.py
running bin/weewx/tests/test_series.py
running bin/weedb/tests/test_errors.py
running bin/weedb/tests/test_weedb.py
running bin/weeplot/tests/test_utilities.py
no failures
no tests were skipped
see build/test-results for output from the tests
gary@stretch19:~/src/weewx$
```

tkeffer commented:
I'm getting the error pretty consistently. On my development machine, using Python 2.7, and 3 different versions of Python 3, and on my Amazon AWS instance using Python 3.6.9

See my earlier post for how I invoked the test, and the results.

gjr80 commented:
hmm, looked but did not see.

I just manually called `scale()` from the python console on pycharm and received the following:
```
>>> import weeplot.utilities
>>> weeplot.utilities.scale(10.0, 10.0+1e-8)
(10.0, 10.1, 0.01)
```
which is consistent with your results. That was python 3.7.3. I checked the results of the `make test` that passed (run on a  Stretch VM) that were error free and it says it used python3, running python3 from the console gives me python 3.5.3.

Found a Buster VM, pulled master and ran `make test` and it errored out. So I guess something has changed python wise since 3.5.3. Will see what I can find.

gjr80 commented:
Hang on a sec, doctest in `utilities.py` shows:
```
    >>> print("(%.2f, %.2f, %.2f)" % scale(10.0, 10.0+1e-8))
    (10.00, 10.10, 0.01)
```
the `test_utilities.py` line that errors out is:
```
        self.assertEqual("(%.5f, %.5f, %.5f)" % scale(10.0, 10.0 + 1e-8),
                         "(10.00000, 10.12000, 0.02000)")
```
Same test but different results expected. Is this a case of an error in the test suite expected results? 

Long story short, I wonder if my PR changed the output of `scale(10.0, 10.0 + 1e-8)` but I only updated the test suite with expected results for a couple of other tests. Somehow when I ran the test suite I ran the test suite on the old code and it passed. The new code was committed, well after 4.1.1 was released so no testing by you or Matthew caught the error until now.

Are we happy that `scale(10.0, 10+1e-8)` returns `(10.0, 10.1, 0.01)` rather than `(10.0, 10.12, 0.02)`? Since that result has been in the doctest for some years perhaps we are?

tkeffer commented:
It looks like that line was changed by me from `(10.0, 10.10, 0.01)` to `(10.0, 10.12, 0.02)` in commit https://github.com/weewx/weewx/commit/98bde920482cb6f69d86b274c29cf623bd5c6285.  

I actually prefer `(10.0, 10.10, 0.01)`. So, can you just change the test suite?

gjr80 commented:
Yes, I think we can just change the test suite. Was a little hesitant at first as I wasn't sure if 10 steps would be too crowded, but looking at a few plots it's not a problem.

I will make it so.

gjr80 commented:
Fixed at commit 533dedc43a32916f5b992a8d4b0af41eec32ec05

tkeffer commented:
Thanks, Gary. I merged `development` into `master` so the fix will appear there. Be sure to do a pull, or you'll be in merge hell.
